crypto-core/src/main/java/com/palantir/crypto2/io/CryptoStreamFactory.java

```diff
@@ -150,4 +151,53 @@ public void close() throws IOException {
             input.close();
         }
     }
+
+    /**
+     * {@link ChunkingOutputStream} limits the size of individual writes to the wrapped {@link OutputStream}
+     * in order to prevent degraded performance on large buffers as described in
+     * <a href="https://github.com/palantir/hadoop-crypto/pull/586">hadoop-crypto#586</a>.
+     */
+    static final class ChunkingOutputStream extends FilterOutputStream {
```

Comment:
any reason a `BufferedOutputStream` won't work for you here?

Reply:
Yep! BufferedOutputStream actually does the opposite of what we want here! Given a large input (beyond the configured buffer size), the BufferedOutputStream will flush any data in its buffer, and write the input buffer directly to the delegate stream as is.

In most cases a BufferedOutputStream is helpful to avoid native overhead for crypto, but I think that's a bit outside of the scope of this change, and introducing a buffered stream around our chunking stream may result in additional unnecessary copies.